### ISSUE TITLE
RFC: Improved implementation of KeepAlive

### ISSUE THREAD

**Issue by jjmaldonis** [deepgram/deepgram-python-sdk#144]
Currently, if a user wants to use Deepgram's KeepAlive functionality for their streaming requests, they must send the KeepAlive message "manually" by making a call to `liveClient.keep_alive()`. This requires an additional thread that 1) watches when messages have been sent over the websocket and 2) uses a timer to keep track of the ~10 second interval before it needs to send a KeepAlive message. This logic is not straightforward to implement, especially for new programmers who may not have a deep understanding of websockets, multithreading, and callbacks / event handlers (for the timer).

We can provide a better user experience, and this RFC issue is a place to discuss.

There is essentially only one way a client would ever want to use KeepAlive messages: A KeepAlive message should be sent if a message has not been sent after X seconds (where X ranges between ~ 3-10 seconds). Since the use case is one-dimensional, we can implement the necessary logic in our SDKs rather than require every use to implement their own version of this logic.

Below is a proposed version of the SDK's interface, using the example in our [readme](https://github.com/deepgram/deepgram-python-sdk#transcribe-audio-in-real-time-or-streaming-audio) as a starting point.

```python
from deepgram import Deepgram
import asyncio
import aiohttp

# Your Deepgram API Key
DEEPGRAM_API_KEY = 'YOUR_API_KEY'

# URL for the audio you would like to stream
URL = 'http://stream.live.vc.bbcmedia.co.uk/bbc_world_service'

async def main():
  # Initialize the Deepgram SDK
  deepgram = Deepgram(DEEPGRAM_API_KEY, keep_alive=10)  # <---- This is the only line that changes!

  # Create a websocket connection to Deepgram
  # In this example, punctuation is turned on, interim results are turned off, and language is set to US English.
  try:
    deepgramLive = await deepgram.transcription.live({ 'punctuate': True, 'interim_results': False, 'language': 'en-US' })
  except Exception as e:
    print(f'Could not open socket: {e}')
    return

# Listen for the connection to close
  deepgramLive.registerHandler(deepgramLive.event.CLOSE, lambda c: print(f'Connection closed with code {c}.'))

  # Listen for any transcripts received from Deepgram and write them to the console
  deepgramLive.registerHandler(deepgramLive.event.TRANSCRIPT_RECEIVED, print)

  # Listen for the connection to open and send streaming audio from the URL to Deepgram
  async with aiohttp.ClientSession() as session:
    async with session.get(URL) as audio:
      while True:
        data = await audio.content.readany()
        deepgramLive.send(data)

        # If there's no data coming from the livestream then break out of the loop
        if not data:
            break

  # Indicate that we've finished sending data by sending the customary zero-byte message to the Deepgram streaming endpoint, and wait until we get back the final summary metadata object
  await deepgramLive.finish()

asyncio.run(main())
```

The most important thing is that the user can implement all the logic for "send a keep alive every 10 seconds if no audio has been sent" _with a single line of code_! This is a much cleaner way for a user to "implement" the KeepAlive functionality (compared to the implementation described in the first paragraph of this PR).

An aside: Technically, the user does not have access to the information to know when a message is sent over the websocket because the Deepgram SDK puts messages in a message queue, which means the message could be sent later than the user expects. If the user assumes the audio data is sent as soon they queue it, and they start a timer at that point, the timer will technically be off. The amount that the timer is off is likely to be insignificant, but in rare circumstances the timer could get out of sync and it's theoretically possible for the websocket to close despite the user sending KeepAlive messages at (what they think) is the appropriate time. This scenario is due to the SDK hiding information from the user about when messages are truly sent to Deepgram. If we were to implement the KeepAlive logic inside the SDKs, we could bypass this hurdle because we have access to the SDK's full set of internal behavior/information.

Pros:
- Users do not need to implement their own KeepAlive behavior, which reduces implementation errors, reduces frustration, and improves users' time-to-value.
- Whenever a user raises concerns about KeepAlive messages and they are using one of our SDKs, Deepgram's engineers will be able to quickly narrow in on the issue and provide helpful feedback faster.

Cons:
- The implementation is non-trivial, so Deepgram will need to spend time to update the KeepAlive logic in each SDK and maintain it. However, currently our customers need to do this themselves (resulting in many times the effort).

**Comment by lukeocodes** [comment]
My comments in support of this RFC:

> I don't think we need to give them the functionality of making "KeepAlive" requests

The SDK should know how to send a KeepAlive, and how long it has been since the last message was sent. The API provides the functionality and the SDK should implement using it while a websocket is open. So the user just "connects" using the SDK and then can "send" audio. This really bakes in the simplicity that I would hope for from an SDK, and abstracts the complexity of the API away from them.

It's a similar maturity step to our work on error handling, and possibly later work on rate limit back offs.

**Comment by jkroll-deepgram** [comment]
I agree that it makes sense for the SDK to have this baked in, as an opt-in functionality. Better for us to take on the implementation and maintenance burden than have each individual customer attempt to do so.

Small point on the queue timing - that must be partially why we recommend to "continue sending KeepAlive messages 3-5 seconds before the 12 second timeout window expires" ([docs](https://developers.deepgram.com/reference/streaming#stream-keepalive)).

Does the customer benefit from being able to specify a frequency, i.e. `keep_alive=10` versus `keep_alive=true`? We could just send the KeepAlive message every ~7 seconds if they opt into the feature, without exposing an integer value.

**Comment by shirgoldbird** [comment]
I'm mulling over `keep_alive=10` vs `keep_alive=true`...

Benefits of specifying an integer:

* Requires user to think more closely about their logic and audio
* Possibly prevents wasting resources on our end/surprise charges on theirs (we don't charge unless audio is being sent, but I can imagine some situation where users might keep a connection open longer than they need to and accidentally just keep sending audio - imagine a script left running in the background?)

Benefits of specifying `true`:

* Most abstraction
* Least error-prone/"why did my connection time out"
* Users might just set a low timeout anyway and effectively get the same behavior

No strong opinion yet, throwing this out for consideration.

**Comment by jjmaldonis** [comment]
The only reason I can think of to use `keep_alive=[int]` rather than `keep_alive=true` is if the user has a bad internet connection.

At the end of the day, the user just wants to tell their code to keep the websocket open until they manually close it, which means `keep_alive=true` makes the most sense.

However, if a user has a very slow internet connection, the keep alive message may take a few seconds to send. So customers with a consistently fast internet connection might want to use `keep_alive=10`, and customers with a slow connection may want to use `keep_alive=3`. Other than this edge case, I can't think of a reason to use `keep_alive=[int]`.

I have a slight preference for `keep_alive=true` to reduce complexity, even at the cost of a poorer experience for very slow internet connections. Realistically, I'm guessing streaming is unlikely to be used on slow internet connections.

**Comment by lukeocodes** [comment]
Another consideration for me; 

In the new architecture, we have no concept of a queue. The requests from the user code, via the SDK, are formatted and sent straight away.

I'm trying to understand the need for a message queue.

But, affectively you'd store the timestamp a message was sent, whether it was from a queue or direct, and fill in when a threshold of delay between messages sent is reached.

I don't think we should make the keep_alive configurable, as it makes no sense to tune it. It is an invisible goal as we don't document the internal function of the API which causes the connection to drop. The keep alive should probably be around `8` seconds to ensure it isn't unnecessarily making calls, and isn't waiting to close to the timeout to react.
